editor/utils/allow-list.js

```diff
@@ -0,0 +1,67 @@
+/**
+ * External dependencies
+ */
+import {
+	castArray,
+	isArray,
+	isPlainObject,
+	isString,
+	mapValues,
+	reduce,
+} from 'lodash';
+
+/**
+ * Parses an allow list shorthand into an allow list object.
+ *
+ * An allow list obejct maps block types to whether or not they are explicitly
+ * allowed as a parent or child. A wildcard ('*') case determines whether or
+ * not parents or children are allowed *in general*.
+ *
+ * Some examples:
+ *
+ * { 'core/image': true, '*': false } specifies that only images are allowed.
+ * { 'core/verse': false, '*': true } specifies that everything except verses are allowed.
```

Comment:
Do we need to support the block lists? I'm asking cause it seems redundant with the `allowedBlocks` prop of the `InnerBlock` component. and in that case we could simplify the API here and just provider a list of allowed parents.

Reply:
Do you mean allow the advanced syntax (`{ 'foo': false, 'bar': true, '*': true }`) in `allowedBlocks` but only simple arrays (`[ 'foo', 'bar' ]`) in `allowedParents`?

My preference is that we support the same syntax in `allowed_block_types` (the PHP filter), `allowedBlocks` (the `InnerBlocks` prop), and `allowedParents` since it's no extra work for us and introduces some consistency throughout the entire blocks API.

Reply:
Oh sorry, what I meant is that we should only support `[ 'foo', 'bar' ]` in all of these, as it's simple and I believe we can achieve everything with it? unless I'm missing a use-case?

Reply:
Doing exceptions is harder with the array-only list (how do you make `({ 'foo': false, 'bar': true, '*': true })`?), so unless we keep the `['!foo', 'bar', '*']` option around (which is more code/complexity and less understandable as plain JS) we'd need to allow objects as arguments.

Reply:
Do we really need to do exceptions? What use-cases can't be achieved with a combination of `allowParents` and `allowedBlocks` (which is `allowChildren`)?

Reply:
Actually we need to cover two cases which are:
- list all allowed blocks
- list all disallowed blocks, which is in other words, all but listed blocks

Reply:
> What use-cases can't be achieved with a combination of allowParents and allowedBlocks (which is allowChildren)?
 
A plugin developer wouldn't be able to e.g. create a block that allows all children except for `core/image` because they can't change the `allowedParents` of an already registered block.

> Do we really need to do exceptions?

I mean—no 😛 But my thinking was that it could be useful for plugin developers.

In the interest of arriving at a consensus I'd be happy to start with only whitelist arrays. We can expand to allowing whitelist+blacklist objects in the future based on feedback we get from plugin developers.

Let me know what y'all think.

Reply:
I think you convinced me we need both, particularly this argument 👍 

> A plugin developer wouldn't be able to e.g. create a block that allows all children except for core/image because they can't change the allowedParents of an already registered block.